### PR TITLE
support serialize and deserialize for engines

### PR DESCRIPTION
**Pull request by RyanKung**

### CODE REVIEW

**Comment by RyanKung** [comment]
@microsoft-github-policy-service agree [company="{Rings Network}"]

**Comment by RyanKung** [comment]
@microsoft-github-policy-service agree company="rings network"

**Comment by srinathsetty** [comment]
Hi @RyanKung Could you please clarify the need for serde on Engine related types?

**Comment by RyanKung** [comment]
> Hi @RyanKung Could you please clarify the need for serde on Engine related types?

My use case is like

```rust
use crate::provider::VestaEngine;
use crate::provider::PallasEngine;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Wrap<E1, E2>
where
      E1: Engine<Base = <E2 as Engine>::Scalar>,
      E2: Engine<Base = <E1 as Engine>::Scalar>,
{
    inner: (E1, E2),
}

#[derive(Serialize, Deserialize, Clone)]
enum Example
{
    VestaPallas(Wrap<VestaEngine, PallasEngine>),
    PallasVasta(Wrap<PallasEngine, VestaEngine>),
}
```

This will cause a compile error:

```shell
error[E0277]: the trait bound `VestaEngine: Serialize` is not satisfied
   --> src/lib.rs:855:10
    |
855 | #[derive(Serialize, Deserialize, Clone)]
    |          ^^^^^^^^^ the trait `Serialize` is not implemented for `VestaEngine`
...
858 |     VestaPallas(Wrap<VestaEngine, PallasEngine>),
    |                 ------------------------------- required by a bound introduced by this call
    |
    = help: the following other types implement trait `Serialize`:
              bool
              char
              isize
              i8
              i16
              i32
              i64
              i128
            and 238 others
```

File: src/provider/mod.rs [hunk 1]
```diff
@@ -26,13 +26,15 @@ use crate::{
   traits::Engine,
 };
 use pasta_curves::{pallas, vesta};
+use serde::Deserialize;
```
Comment (srinathsetty) [inline src/provider/mod.rs:29]: Nit: use serde::{Deserialize, Serialize};

File: src/lib.rs [hunk 2]
```diff
@@ -53,7 +53,7 @@ use traits::{
 };
 
 /// A type that holds public parameters of Nova
-#[derive(Serialize, Deserialize)]
+#[derive(Serialize, Deserialize, Clone)]
```
Comment (srinathsetty) [inline src/lib.rs:56]: Why do we need Clone on public parameters? This type is always passed by reference, so clone seems unnecessary.

File: src/provider/mod.rs (hunk 1 above)
Comment (RyanKung) [reply src/provider/mod.rs]: Thanks, just updated.

**Pull request merged**